extensions/druid-rocketmq/src/main/java/io/druid/firehose/rocketmq/RocketMQFirehoseFactory.java

```diff
@@ -0,0 +1,571 @@
+   */
+  private boolean hasMessagesPending()
+  {
+
+    for (ConcurrentHashMap.Entry<MessageQueue, ConcurrentSkipListSet<MessageExt>> entry : messageQueueTreeSetMap.entrySet()) {
+      if (!entry.getValue().isEmpty()) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  @Override
+  public Firehose connect(ByteBufferInputRowParser byteBufferInputRowParser) throws IOException, ParseException
+  {
+
+    Set<String> newDimExclus = Sets.union(
+        byteBufferInputRowParser.getParseSpec().getDimensionsSpec().getDimensionExclusions(),
+        Sets.newHashSet("feed")
+    );
+
+    final ByteBufferInputRowParser theParser = byteBufferInputRowParser.withParseSpec(
+        byteBufferInputRowParser.getParseSpec()
+                                .withDimensionsSpec(
+                                    byteBufferInputRowParser.getParseSpec()
+                                                            .getDimensionsSpec()
+                                                            .withDimensionExclusions(
+                                                                newDimExclus
+                                                            )
+                                )
+    );
+
+    /**
+     * Topic-Queue mapping.
+     */
+    final ConcurrentHashMap<String, Set<MessageQueue>> topicQueueMap;
+
+    /**
+     * Default Pull-style client for RocketMQ.
+     */
+    final DefaultMQPullConsumer defaultMQPullConsumer;
+    final DruidPullMessageService pullMessageService;
+
+    messageQueueTreeSetMap.clear();
+    windows.clear();
+
+    try {
+      defaultMQPullConsumer = new DefaultMQPullConsumer(this.consumerGroup);
+      defaultMQPullConsumer.setMessageModel(MessageModel.CLUSTERING);
+      topicQueueMap = new ConcurrentHashMap<>();
+
+      pullMessageService = new DruidPullMessageService(defaultMQPullConsumer);
+      for (String topic : feed) {
+        Validators.checkTopic(topic);
+        topicQueueMap.put(topic, defaultMQPullConsumer.fetchSubscribeMessageQueues(topic));
+      }
+      DruidMessageQueueListener druidMessageQueueListener =
+          new DruidMessageQueueListener(Sets.newHashSet(feed), topicQueueMap, defaultMQPullConsumer);
+      defaultMQPullConsumer.setMessageQueueListener(druidMessageQueueListener);
+      defaultMQPullConsumer.start();
+      pullMessageService.start();
+    }
+    catch (MQClientException e) {
+      LOGGER.error("Failed to start DefaultMQPullConsumer", e);
+      throw new IOException("Failed to start RocketMQ client", e);
+    }
+
+    return new Firehose()
+    {
+
+      @Override
+      public boolean hasMore()
+      {
+        boolean hasMore = false;
+        DruidPullRequest earliestPullRequest = null;
+
+        for (Map.Entry<String, Set<MessageQueue>> entry : topicQueueMap.entrySet()) {
+          for (MessageQueue messageQueue : entry.getValue()) {
+            if (messageQueueTreeSetMap.keySet().contains(messageQueue)
```

Comment:
As you noted, I do not synchronize there to make the whole looping theoretically thread-safe. The fact is, we do not need it.  This loop intends to figure out if we need to send pull-requests to ``DruidPullMessageService``, snapshot states of the collection would suffice and it won't matter in corner case that we sends **more** pull-requests. Note only read operation is involved.

Real updates to the concurrent collection are made in ``DruidPullMessageService``, thread-safety of ``ConcurrentHashMap`` suffices:
``
if (!messageQueueTreeSetMap.keySet().contains(pullRequest.getMessageQueue())) {
                messageQueueTreeSetMap.putIfAbsent(
                    pullRequest.getMessageQueue(),
                    new ConcurrentSkipListSet<>(new MessageComparator())
                );
              }
              messageQueueTreeSetMap.get(pullRequest.getMessageQueue()).addAll(pullResult.getMsgFoundList());
``